File: csharp/854.k-similar-strings.cs
/*
 * @lc app=leetcode id=854 lang=csharp
 *
 * [854] K-Similar Strings
 */

using System.Collections.Generic;

// @lc code=start
public class Solution {
    public int KSimilarity(string A, string B) {
        if (A == B) {
            return 0;
        } 
        var visited = new HashSet<string>();
        var queue = new Queue<string>();
        int ret = 1, n = A.Length;
        visited.Add(A);
        queue.Enqueue(A);
        while (true) {
            for (int k = queue.Count; k > 0; --k) {
                var curr = queue.Dequeue();
                var arr = curr.ToCharArray();
                int i = 0;
                while (arr[i] == B[i]) {
                    ++i;
                }
                char c = arr[i], tc = B[i];
                for (int j = i + 1; j < n; ++j) {
                    if (arr[j] == B[j] || arr[j] != tc) {
                        continue;
                    }
                    arr[i] = arr[j];
                    arr[j] = c;
                    var next = new string(arr);
                    if (next == B) {
                        return ret;
                    }
                    if (visited.Add(next)) {
                        queue.Enqueue(next);
                    }
                    arr[j] = arr[i];
                }
            }
            ++ret;
        }
        return -1;
    }
}
// @lc code=end

File: java/854.k-similar-strings.java
/*
 * @lc app=leetcode id=854 lang=java
 *
 * [854] K-Similar Strings
 */

// @lc code=start
class Solution {
    public int kSimilarity(String A, String B) {
        Set<String> visited = new HashSet<String>();
        Queue<String> queue = new LinkedList<String>();
        int ret = 0, n = A.length();
        visited.add(A);
        queue.offer(A);
        while (queue.size() > 0) {
            for (int i = queue.size(); i > 0; --i) {
                String curr = queue.poll();
                if (curr.equals(B)) {
                    return ret;
                }
                int l = 0;
                while (curr.charAt(l) == B.charAt(l)) {
                    ++l;
                }
                for (int r = l + 1; r < n; ++r) {
                    char c = curr.charAt(r);
                    if (c == B.charAt(r) || c != B.charAt(l)) {
                        continue;
                    }
                    StringBuilder sb = new StringBuilder(curr);
                    sb.replace(r, r + 1, Character.toString(curr.charAt(l)));
                    sb.replace(l, l + 1, Character.toString(c));
                    String next = sb.toString();
                    if (visited.add(next)) {
                        queue.offer(next);
                    }
                }
            }
            ++ret;
        }
        return -1;
    }
}
// @lc code=end

